### PR TITLE
allow overriding of DWN hosts

### PR DESCRIPTION
**Pull request by michaelneale**
adds this enhancement: https://github.com/TBD54566975/web5-js/issues/86

### CODE REVIEW

**Comment by csuwildcat** [comment]
Would rather call the property dwnEndpoints, as host makes it seem like an external entity is always involved, when it could simply be a static IP to a box in your own home.

**Comment by michaelneale** [comment]
@csuwildcat yeah I agree - was struggling for a name. endpoints is also what DID spec calls them, will change.

**Comment by michaelneale** [comment]
@csuwildcat how does this look?

**Comment by csuwildcat** [comment]
Lgtm

**Comment by frankhinek** [comment]
@mistermoe were you going to push up some additions to the PR we discussed yesterday to make overriding other properties possible?  Or should we start with `dwnEndpoints` and open a separate PR?

**Comment by mistermoe** [comment]
i didn't get a chance to add those additional overrides we discussed yesterday. let's start with `dwnEndpoints`. i'll open a separate PR that introduce the other options

**Comment by mistermoe** [comment]
the only thing that gives me pause at the moment is that this `dwnEndpoints` override option only makes sense in a scenario where an embedded user agent is being used. That said, the only agent that exists today is an embedded agent haha! the long term goal is for desktop & mobile agents to exist and it's in those external agents where you'd be providing your custom dwn endpoints or any other custom did configurations. that wouldn't happen via the `web5` api.

Given that we haven't yet finished the first external agent, i think there's value in allowing people to provide custom dwn endpoints today. I think we could do it like so:

```javascript
const { web5, did } = await Web5.connect({
  techPreview: {
    dwnEndpoints: []
 }
});
```

namespacing the `dwnEndpoints` option within a `techPreview` object hopefully makes it more clear that the option is temporary

File: README.md [hunk 1]
```diff
@@ -60,6 +60,13 @@ Enables an app to request connection to a user's local identity app (like a desk
 const { web5, did: myDid } = await Web5.connect();
 ```
 
+connect() takes in an optional `options` object with the following properties: `dwnHosts` which is a list of DWeb Node endpoints to connect to if a new DID is to be created. If no `dwnHosts` are provided, the SDK will default to built in bootstrap nodes.
+
+For example:
+```javascript
+const { web5, did: myDid } = await connect({ 'dwnHosts': ['https://dwn.tbddev.org/dwn0'] });
```
Comment (frankhinek) [inline README.md]: Should be `Web5.connect()` and not `connect()`

**Comment by frankhinek** [comment]
Spotted a couple issues with the existing PR:
- only 1 of the 2 README files was updated
- README example was `const { web5, did: myDid } = await connect(...` instead of `const { web5, did: myDid } = await Web5.connect(...`.

both resolved with Commit [05e4e34](https://github.com/TBD54566975/web5-js/pull/93/commits/05e4e34c0535dda087913e20ea1a294fb2013312)

Still one remaining issue which is if `dwnEndpoints` is an empty array, the `messageAuthorizationKeys` and `recordEncryptionKeys` values aren't populated in the DID document:

with `{ techPreview: { dwnEndpoints: [] } }`:

Screenshot 2023-05-26 at 12 52 46 PM

with `{ techPreview: { dwnEndpoints: ['https://dwn.your-domain.org'] } }`:

Screenshot 2023-05-26 at 12 55 53 PM

Discussing with @mistermoe and @csuwildcat as to how we should handle.

File: packages/dids/tests/tech-preview.spec.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import type { DwnServiceEndpoint } from '../src/types.js';
+
+import { expect } from 'chai';
+
+import { DidIonApi } from '../src/did-ion.js';
+
+describe('Tech Preview', function () {
+  describe('generateDwnConfiguration()', () => {
+    it.only('returns keys and services with two DWN URLs', async () => {
```
Comment (diehuxx) [inline packages/dids/tests/tech-preview.spec.ts:9]: what's with the `it.only`s? are you intentionally trying to exclude other tests from the `package/dids/test` suite? If so, can you use `.skip` on the ones you want to skip instead?

Comment (frankhinek) [reply packages/dids/tests/tech-preview.spec.ts]: Thanks for catching.  Should have been removed.

File: packages/web5/src/web5.ts [hunk 3]
```diff
@@ -110,20 +112,25 @@ export class Web5 {
   }
 
   /**
-   * dynamically selects up to 4 dwn hosts
+   * Dynamically selects up to 2 DWN endpoints that are provided
+   * by default during the Tech Preview period.
    */
-  static async getDwnHosts() {
+  static async getTechPreviewDwnEndpoints(): Promise<string[]> {
     const response = await fetch('https://dwn.tbddev.org/.well-known/did.json');
```
Comment (mistermoe) [inline packages/web5/src/web5.ts:119]:
```suggestion
    let response;

    try {
      response = await fetch('https://dwn.tbddev.org/.well-known/did.json');
    } catch(e) {
      console.warn('failed to get tech preview dwn endpoints', e);
      return [];
    }
```

File: packages/web5/src/web5.ts [hunk 4]
```diff
@@ -110,20 +112,25 @@ export class Web5 {
...
 
     // allocate up to 2 nodes for a user.
-    const numNodesToAllocate = Math.min(Math.floor(nodes.length / 2), 2);
-    const dwnUrls = new Set([]);
+    const dwnUrls = new Set<string>();
+    let attempts = 0;
+    const numNodesToAllocate = Math.min(nodes.length, 2);
 
-    for (let i = 0; i < numNodesToAllocate; i += 1) {
+    while(dwnUrls.size < numNodesToAllocate && attempts < nodes.length) {
```
Comment (mistermoe) [inline packages/web5/src/web5.ts:133]: thoughts on the following?

```typescript
// allocate up to 2 nodes for a user.
const dwnUrls = new Set<string>();
const numNodesToAllocate = Math.min(nodes.length, 2);

for (let attempts = 0; attempts < nodes.length && dwnUrls.size < numNodesToAllocate; attempts += 1) {
  const nodeIdx = getRandomInt(0, nodes.length);
  const dwnUrl = nodes[nodeIdx];

  try {
    const healthCheck = await fetch(`${dwnUrl}/health`);
    if (healthCheck.status === 200) {
      dwnUrls.add(dwnUrl);
    }
  } catch(e: unknown) {
    // Ignore healthcheck failures and try the next node.
  }
}
```

smushes `attempts`

**Comment by michaelneale** [comment]
yeah calling connect with dwnEndpoints specified is in some ways assuming that is the special case that a did will be created for you, but wasn't sure where else it belongs in the short term.

File: packages/web5/src/web5.ts (hunk 3 above)
Comment (frankhinek) [reply packages/web5/src/web5.ts:119]: Incorporated into latest commit.

File: packages/web5/src/web5.ts (hunk 4 above)
Comment (frankhinek) [reply packages/web5/src/web5.ts:133]: kinda doesn't smush `attempts` since its still there in the `for` loop 😄 
but LGTM

**Review by mistermoe — APPROVED** [review]
🥇

**Pull request merged**